### PR TITLE
Pulse AVL tree Implementation

### PR DESCRIPTION
**Pull request by sheeraSearch82**
Pulse AVL tree implementation and interface. Spec is inspired from Steel.

### CODE REVIEW

File: lib/Pulse.Lib.AVLTree.fsti [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+module Pulse.Lib.AVLTree
+open Pulse.Lib.Pervasives
+module S = Pulse.Lib.Stick.Util
+module FA = Pulse.Lib.Forall.Util
+open FStar.List.Tot
+
+module T = Spec.AVLTree
+module G = FStar.Ghost
+
+noeq
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fsti:12]: As far as I can see you can hide node, node_ptr from the interface and make the implementation of tree_t hidden from the interface.

File: lib/Pulse.Lib.AVLTree.fsti [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  (ensures fun b -> is_tree x ft ** pure (b <==> (ft == T.Leaf)))
+
+val null_tree_t (t:Type0) : tree_t t
+
+val create (t:Type0)
+   : stt (tree_t t)
+  (requires emp)
+  (ensures fun x -> is_tree x T.Leaf)
+
+val append_left_none (#t:Type0) (x:tree_t t) (v:t) (#ft:G.erased (T.tree t))
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fsti:46]: Do we need the appends, rotates, balalance etc. functions in the interface? Can't the interface just have create, insert, delete, mem, and height?

File: lib/Spec.AVLTree.fst [hunk 3]
```diff
@@ -0,0 +1,592 @@
+module Spec.AVLTree
```
Comment (aseemr) [inline lib/Spec.AVLTree.fst:1]: Add the copyright headers in all the files, and add yourself as author. See Pulse.Lib.FractionalAnchoredPreorder.fst in lib for an example.

Comment (aseemr) [inline lib/Spec.AVLTree.fst:1]: For this file specifically, please add a comment at the top that this file is copied verbatim from the steel repository.

File: lib/Pulse.Lib.AVLTree.fst [hunk 4]
```diff
@@ -0,0 +1,948 @@
...
+//----------------------------------------------------------------------------------------------------------
+module Pulse.Lib.AVLTree
+open Pulse.Lib.Pervasives
+module S = Pulse.Lib.Stick.Util
+module FA = Pulse.Lib.Forall.Util
+open FStar.List.Tot
+
+module T = Spec.AVLTree
+
+//type vprop = slprop
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:11]: Delete?

File: lib/Pulse.Lib.AVLTree.fst [hunk 5]
```diff
@@ -0,0 +1,948 @@
...
+let rec is_tree #t ct ft = match ft with
+  | T.Leaf -> pure (ct == None)
+  | T.Node data left' right' ->
+    exists* (p:node_ptr t) (lct:tree_t t) (rct:tree_t t).
+      pure (ct == Some p) **
+      pts_to p { data = data ; left = lct ; right = rct} **
+      is_tree lct left' **
+      is_tree rct right'
+
+//boilerplate$
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:22]: There are many such $ ending comments in the file, can we delete them?

File: lib/Pulse.Lib.AVLTree.fst [hunk 6]
```diff
@@ -0,0 +1,948 @@
...
+
+```pulse
+ghost
+fn elim_is_tree_leaf (#t:Type0) (x:tree_t t)
+requires is_tree x T.Leaf 
+ensures pure (x == None)                 
+{
+   
+   unfold (is_tree x T.Leaf) 
+
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:32]: Redundant blank lines?

File: lib/Pulse.Lib.AVLTree.fsti [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+val is_tree #t (ct:tree_t t) (ft:T.tree t)
+: Tot vprop (decreases ft)
+
+val height (#t:Type0) (x:tree_t t) (#ft:G.erased (T.tree t))
+  : stt nat
+(requires is_tree x ft)
+(ensures fun n -> is_tree x ft ** pure (n == T.height ft))
+
+val is_empty (#t:Type) (x:tree_t t) (#ft:(T.tree t))
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fsti:34]: The spec tree should be erased.

File: lib/Pulse.Lib.AVLTree.fsti [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+   : stt (tree_t t)
+(requires is_tree x ft)
+(ensures fun y -> is_tree y  (T.append_right ft v))
+
+val node_data (#t:Type) (x:tree_t t) (#ft:G.erased (T.tree t))
+     : stt t
+(requires is_tree x ft  ** (pure (Some? x)))
+(ensures fun _ -> is_tree x ft)
+
+val is_mem (#t:eqtype) (x:tree_t t) (v: t) (#ft:G.erased (T.tree t))
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fsti:66]: Nit, we can just call it mem.

File: lib/Pulse.Lib.AVLTree.fst [hunk 9]
```diff
@@ -0,0 +1,948 @@
...
+      } else {
+        (r_height + 1)
+      }
+    }
+  }
+}
+```
+
+```pulse
+fn add_k_height (#t:Type0) (x:tree_t t) (k:nat)
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:173]: What is this add_k_height? Can we remove it?

File: lib/Pulse.Lib.AVLTree.fst [hunk 10]
```diff
@@ -0,0 +1,948 @@
...
+ensures  is_tree x ft ** pure (T.Node? ft)
+{
+  cases_of_is_tree (Some v) ft;
+  unfold is_tree_cases;
+  intro_is_tree_node (Some v) v;
+}
+
+```
+
+//ensures is_tree x 'l ** pure (n == T.height 'l)
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:806]: This line and the next one can go?

File: lib/Pulse.Lib.AVLTree.fst [hunk 11]
```diff
@@ -0,0 +1,948 @@
+//Pulse AVL tree implementation. Inspired from Steel. The FStar spec file is adopted from Steel
+//----------------------------------------------------------------------------------------------------------
+module Pulse.Lib.AVLTree
+open Pulse.Lib.Pervasives
+module S = Pulse.Lib.Stick.Util
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:5]: This import and the next one are not used in the module as far as I can see.

**Comment by aseemr** [comment]
Hi Sheera, the files should be in lib/pulse/lib directory, with all other Pulse libraries.

File: lib/Pulse.Lib.AVLTree.fsti [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    left : tree_t t;
+    right : tree_t t;
+}
+
+and node_ptr (t:Type0) = ref (node t)
+
+//A nullable pointer to a node
+and tree_t (t:Type0) = option (node_ptr t)
+
+let vprop = slprop
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fsti:24]: This should go away.

File: lib/Pulse.Lib.AVLTree.fst [hunk 13]
```diff
@@ -0,0 +1,948 @@
...
+              intro_is_tree_node x vl;
+              b3;
+              
+            }
+        }
+    }
+}
+```
+
+module G = FStar.Ghost
```
Comment (aseemr) [inline lib/Pulse.Lib.AVLTree.fst:423]: There are multiple of this abbreviation in this file.

File: lib/Pulse.Lib.AVLTree.fsti (hunk 1 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fsti]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fsti (hunk 2 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fsti]: comment addressed.

File: lib/Spec.AVLTree.fst (hunk 3 above)
Comment (sheeraSearch82) [reply lib/Spec.AVLTree.fst]: comment addressed.

Comment (sheeraSearch82) [reply lib/Spec.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 4 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 5 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 6 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fsti (hunk 7 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fsti]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fsti (hunk 8 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fsti]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 9 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst [hunk 14]
```diff
@@ -0,0 +1,948 @@
...
+    Some v -> {
+      is_tree_case_some x v;
+      v
+    }
+  }
+}
+```
+
+```pulse
+fn rotate_left (#t:Type0) (tree:tree_t t) (#l: G.erased (T.tree t){ Some? (T.rotate_left l) })
```
[the remark this answers is not in the document]
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 10 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 11 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fsti (hunk 12 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fsti]: comment addressed.

File: lib/Pulse.Lib.AVLTree.fst (hunk 13 above)
Comment (sheeraSearch82) [reply lib/Pulse.Lib.AVLTree.fst]: comment addressed.

**Pull request merged**